### ISSUE TITLE
Static IP configuration screws Gateway settings? 

### ISSUE THREAD

**Issue by mreferre** [vmware/photon#402]
I have just try to reconfigure a default PhotonOS (full) install to change from DHCP to Static. 

If I login into my (default) DHCP configuration I can ping 8.8.8.8 
The `ip route` command shows: 

       default via 192.168.1.0/24 dev eno16780032 proto dhcp metric 1024
       192.168.1.0/24 dev eno16780032 proto kernel scope link src 192.168.1.245
       192.168.1.254 dev eno16780032 proto dhcp scope link metric 1024

I then try to change the configuration to static. I mv the `/etc/systemd/network/10-dhcp-en.network` file into `/etc/systemd/network/10-static-en.network` and then I edit it as follows:

       Name=*
 
       [Network]
       Address=192.168.1.44/24
       Gateway=192.168.1.254

I reboot the VM and I see that it picks up the `.44` static address (and I can ping all other hosts in the same subnet) but if i try to ping 8.8.8.8 it gives me a black eye (as if it doesn't route out of the local subnet any more). 

The `ip route` command shows:

        default via 192.168.1.254 dev eno16780032 proto static
        default via 192.168.1.254 dev lo proto static
        192.168.1.0/24 dev eno16780032 proto kernel scope link src 192.168.1.44 

Is there anything else that needs to be done?

**Comment by vinaykul** [comment]
I'm unable to repro the issue in my setup, however "Name=*" may not work very well for static IP. If you don't have more than one interface, could you please try setting Name=eno1 (the default interface name)?

**Comment by mreferre** [comment]
Thanks. 
I have retried (on Fusion) what I was doing yesterday (on vSphere) and it worked. 
Note I have not yet enabled Docker on my instance so I only have one interface (enoxxxxxxx). So the "*" worked just fine. 
The moment I have enabled Docker the 172.16.42 interface came up and having "Name=*" in the configuration file set both at the same static IP (which is obviously NOT what you want). So I had to put "Name=eno1*" so it would pick the host interface (note that using "Name=eno" as you suggested didn't work, most likely it needs the "*" wild card to be able to match the correct name?). 

Long story short (net net) is that the procedure above works with Fusion but NOT on vSphere (for some odd reasons). I don't have access to my vSphere lab right now but I will triple check tomorrow that I was doing the same thing. 

Are you by any chance trying to replicate it on Fusion? Can you try on vSphere?

**Comment by FabioChiodini** [comment]
Hi

i am trying this on vSphere deploying the OVA photon-1.0TP2.ova.

I've tried with:

Name=*
Name=eth0
Name=eno1
Name=en1

With no luck.

Best option is with equal *: it gets the address but ignores the default gateway settings.

Thank you very much.

F.

**Comment by mreferre** [comment]
Fabio, I think that if you try with "Name=eno1*" that will set the IP on the proper interface (and not on the Docker0 internal switch). However the conclusion I got to is that (on vSphere) even if you are able to set the proper IP, the Gateway seems to be getting ignored.

**Comment by vinaykul** [comment]
Hi,

systemd-networkd provides examples of how to configure static addresses for multiple interfaces. Please see https://wiki.archlinux.org/index.php/Systemd-networkd . You'll notice they are not using wildcards.

AFAIK, The 'Name=*' tells systemd to match all interface names and apply the specified properties to it (static address in this case). That's the reason I suspected it was not working correctly for your case - I see default route over the 'lo' interface, and if your route-lookup is picking that up (mine didn't), you wont get anywhere.

I'm used Fusion when I tried this out earlier. By vSphere, do you mean photon running on ESX that you are managing through vSphere client?

Thanks,
Vinay

**Comment by FabioChiodini** [comment]
Thanks Vinaykul.

This is photon installed on an ESXi server.

Following Massimo's advice:

Boot up a new .ova
Used name=eno1* in the config file
Rebooted VM (and the network was ok, gateway working)
Enabled docker

Is this validated or shall I avoid the * notation and write the full name of the interface? ;)

Thank you very much.

F.

**Comment by mreferre** [comment]
I believe what happens (I think) is as follows: 
- if you set name=* the eno1 interface gets changed but the gateway / routing may get assigned to loopback as you pointed out (thus making routing not working)
- if you set name=eno1 it won't recognize what eno1 is and will fail to configure the network (it won't appear with ifconfig) 
- if you set name=eno1* everything will work fine (the static IP gets assigned to the interface as well as the routing rule is bound to it) 


I believe the reason that in the link you pointed out there is no `*` is because the network they are using as an example (e.g. `enp1s0`, `wlp2s0` etc) are the exact interface names. 

My interface isn't `eno1`, it is `eno16780032` 

Massimo.

**Comment by mreferre** [comment]
Also, wether we are using Fusion or vSphere is irrelevant apparently.

**Comment by vinaykul** [comment]
Aha .. my bad, I'm using a later version of photon where we have turned off the not so predictable "predictable network interface renaming" feature and my default is eno1 instead of eno<some large number>. So eno1 worked for me, but eno1* worked for you.

I recommend specifying the full name of the NIC rather than eno1* even though that gets the job done for now. If you were to add another NIC and it happened to show up as eno1<somethingelse> then you would end up with the same static address on two interfaces in the same routing domain. This can cause problems for BSD socket apps.

About vSphere-ESX, I did see the routing issue on a ESX VM with Name=*. I was picking up lo in that case as I suspected. You can see what route-lookup is doing for a specific dest IP using "ip route get". For e.g. 'ip route get 8.8.8.8'

Vinay

**Comment by FabioChiodini** [comment]
Thank you very much Vinay and Massimo :)
